mesonbuild/compilers/cpp.py

```diff
@@ -642,6 +649,13 @@ def get_option_compile_args(self, options: 'KeyedOptionDictType') -> T.List[str]
 
         if options[key.evolve('debugstl')].value:
             args.append('-D_GLIBCXX_DEBUG=1')
+
+            # We can't do _LIBCPP_DEBUG because it's unreliable unless libc++ was built with it too:
+            # https://discourse.llvm.org/t/building-a-program-with-d-libcpp-debug-1-against-a-libc-that-is-not-itself-built-with-that-define/59176/3
+            # Note that unlike _GLIBCXX_DEBUG, _MODE_DEBUG doesn't break ABI. It's just slow.
+            if version_compare(self.version, '>=18'):
+                args.append('-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG')
+
```

Comment:
This is being implemented on the `icc` compiler, but I think you wanted it on the ClangCPPCompiler class instead or additionally (plus inherited by AppleClang).

That's where you handled D_GLIBCXX_DEBUG up above.

Reply:
Hmm, looks like this wasn't actually changed (or else it got reverted in a rebase?)

Reply:
Looking...